src/inline-assembly.md

```diff
@@ -37,515 +50,878 @@ assert_eq!(x, 4 * 6);
+asm := "asm!(" asm_inner ")"
+global_asm := "global_asm!(" asm_inner ")"
+
+non_format_char := ANY_CHAR // except "{" and "}"
+operand_specifier := ident / DEC_LITERAL
+expansion_specifier := *non_format_char
+format_specifier := "{" [operand_specifier] [":" *expansion_specifier]  "}"
+format_escape := "{{" / "}}"
+asm_string_piece := non_format_char / format_specifier / format_escape
+asm_string_content := [*asm_string_piece]
+```
+
+## Invocation [asm.invocation]
+
+r[asm.invocation.asm]
+The [`core::arch::asm!`] macro shall be expanded in an expression context only. The input tokens shall match the `asm_inner` production. The expansion is [`unsafe`][static.expr.safety] and has type `()`, unless the option `noreturn` is specified, in which case it has type `!`.
+
+```rust
+# #[cfg(target_arch = "x86_64")] 
+pub fn main(){
+  unsafe{
+    core::arch::asm!("")
+  }
+}
+```
+
+r[asm.invocation.global_asm]
+The [`core::arch::global_asm!`] macro shall be expanded in an item context only. The input tokens shall match the `asm_inner` production. If the macro is expanded in a function, the program is ill-formed. 
+
+<!--TODO: Test `global_asm!`-->
+
+```rust,ignore
+# #[cfg(target_arch = "x86_64")]
+core::arch::global_asm!(".rodata", "FOO:", ".ascii \"Hello World\"");
+```
+
+```rust,compile_fail
+# #[cfg(target_arch = "x86_64")] 
+pub fn main(){
+    core::arch::global_asm!("FOO:", ".ascii \"Hello World\"");
+}
+```
+
+r[asm.invocation.format-string]
+Each `format_string` input to the [`core::arch::asm!`] and [`core::arch::global_asm!`] macros shall be an expanded string literal for which the content matches the `asm_string_piece` production.
+
+> [!NOTE]
+> an expanded string literal is a string literal (after expanding macros like [`core::concat!`]) that has had every unicode escape sequence replaced with the (appropriately escaped as needed) matching character, and which has been normalized from a raw string literal.
+
+
+r[asm.invocation.concat]
+If multiple `format_string` inputs are provided, then they are concatenated as though by the [`core::concat!`] macro, separating each `format_string` with a string containing a single newline character. If any `format_string` begins a `format_specifier` that is not terminated before the end of the `format_string`, the program is ill-formed. The resulting string is known as the *joined asm-string*
+
+```rust
+# #[cfg(target_arch = "x86_64")] {unsafe{
+let mut x: i32;
+// The following lines are equivalent
+core::arch::asm!("mov rax, 5", "mov rcx, rax", out("rax") x, out("rcx") _); 
+core::arch::asm!("mov rax, 5\nmov rcx, rax", out("rax") x, out("rcx") _);  
+# }}
+```
+
+r[asm.invocation.operands]
+Each operand, other than an explicit register operand ([asm.operands.register]) shall be mentioned by at least one format_specifier in the *joined asm-string*. Explicit registers may not be referred to be a format_specifier.
+
+```rust,compile_fail
```

Comment:
These compile_fail annotations won't work with non-x86_64 architectures (since it won't fail to compile).

Reply:
Yeah, I'll add a
```
# #[cfg(not(target_arch = "x86_64")]
# core::compile_error!("Tests currently require x86_64. Please add per-architecture tests to run off of x86_64");